src/nrniv/nrnbbcore_write.cpp

```diff
@@ -1884,92 +1884,154 @@ static core2nrn_callback_t cnbs[]  = {
   {NULL, NULL}
 };
 
+#if defined(HAVE_DLFCN_H)
+
 extern int nrn_use_fast_imem;
+extern char* neuron_home;
+
+/** Check if coreneuron is loaded into memory */
+bool is_coreneuron_loaded() {
+    bool is_loaded = false;
+    // check if corenrn_version symbole can be found
+    void * handle = dlopen(NULL, RTLD_NOW | RTLD_GLOBAL);
+    if (handle) {
+        void* fn = dlsym(handle, "corenrn_version");
+        is_loaded = fn == NULL ? false : true;
+        dlclose(handle);
+    }
+    return is_loaded;
+}
 
-/** check if file with given path exist */
+/** Check if file with given path exist */
 bool file_exist(const std::string& path) {
-  std::ifstream f(path.c_str());
-  return f.good();
+    std::ifstream f(path.c_str());
+    return f.good();
 }
 
-extern "C" {
-  extern char* neuron_home;
+/** Open library with given path and return dlopen handle **/
+void* get_handle_for_lib(const char* path) {
+    void* handle = dlopen(path, RTLD_NOW|RTLD_GLOBAL);
+    if (!handle) {
+      fputs(dlerror(), stderr);
+      fputs("\n", stderr);
+      hoc_execerror("Could not dlopen CoreNEURON mechanism library : ", path);
+    }
+    return handle;
 }
 
-#if defined(HAVE_DLFCN_H)
+/** Get CoreNEURON mechanism library */
+void* get_coreneuron_handle() {
+	// if already loaded into memory, directly return handle
+	if (is_coreneuron_loaded()) {
+		return dlopen(NULL, RTLD_NOW | RTLD_GLOBAL);
+	}
+
+	// env variable get highest preference
+	const char* corenrn_lib = getenv("CORENEURONLIB");
+	if (corenrn_lib && file_exist(corenrn_lib)) {
+		return get_handle_for_lib(corenrn_lib);
+	}
+
+	// name of coreneuron library based on platform
+	#if defined(MINGW)
+	std::string corenrn_mechlib_name("libcorenrnmech.dll");
+	#elif defined(DARWIN)
+	std::string corenrn_mechlib_name("libcorenrnmech.dylib");
+	#else
+	std::string corenrn_mechlib_name("libcorenrnmech.so");
+	#endif
+
+	// first check if coreneuron specific library exist in <arhc>/.libs
+	std::stringstream s_path;
+	s_path << NRNHOSTCPU << "/.libs/" << corenrn_mechlib_name;
+	std::string path = s_path.str();
+	if (file_exist(path)) {
+		return get_handle_for_lib(path.c_str());
```

Comment:
This isn't a full path.. Should get_handle_for_lib transform to full path. This is also somewhat haphazard in nrnoc/init.c